### DISCUSSION TITLE
Require content collection config

### DISCUSSION THREAD

**Discussion by adamkrupke** [withastro/roadmap#817] [Proposal]
### Body

# Summary

When using content collection API's such as `getCollection()`, do not expose collections that aren't configured.

# Background & Motivation

Today I had a rough session trying to debug why my content integration wasn't working. I had created an `article` collection which I exported from the `src/content/config.ts` file.

```typescript
// src/content/config.ts

import { defineCollection, z } from "astro:content";

const articleCollection = defineCollection({
  type: "content",
  schema: ({ image }) =>
    z.object({
      title: z.string(),
      description: z.string(),
    }),
});

export const collections = {
  article: articleCollection,
};
```

I added some articles to the `src/content/articles` directory. I was happy to see autocomplete for `getCollection("articles")` and I could loop through each item and display its title & description.

<img width="267" alt="image" src="https://github.com/withastro/roadmap/assets/25898698/c0bb0cab-dfa0-43f9-ae00-a98f6bba23e1">

Here I started to notice something was off. The article `data` prop type was `any`. Oh well, maybe it's not typesafe after all. Then I wanted to add a cover image, following [the documentation](https://docs.astro.build/en/guides/images/#images-in-content-collections) closely, but consistently getting the `LocalImageUsedWrongly`:

<img width="1100" alt="image" src="https://github.com/withastro/roadmap/assets/25898698/2aa12b34-6617-494c-b4e9-a3162a603ecd">

After a long while I discovered this was due to my `article` collection being incorrectly named in its singular form. After changing the collection name from `article` to `articles`, typings and the image integration started working.

# Goals

It should've been more obvious something was wrong. I'm thinking any of these actions could be taken:

1. Ideally we shouldn't expose unconfigured collections in the collections API (`getCollection()`)
2. Alternatively, show a runtime warning in development to notify the developer accessing an unconfigured collection.
3. If this neither of above are feasible, then could we at least change the `data` property type from `any` to `unknown`?

**Comment by delucis** [comment]
The current behaviour is kind of intentional as it allows people who don’t care about schemas and types to still use collections they have without setting a schema.

But maybe some kind of logging that warns in cases like this one would be handy, i.e. there’s a collection declared called `X`, but no content for `X` and instead content for a name _similar_ to `X`, then log a warning.

**Reply by adamkrupke** [reply]
That sounds reasonable. I understand the intention. Could an option be to expose both existing content and configured collections, such that I would get autocomplete for both `article` and `articles`? I think that would have me raise an eyebrow and discover the error much sooner.

**Reply by delucis** [reply]
Ah yeah, good point — that might be possible!

cc @bholmesdev

**Reply by bholmesdev** [reply]
Ah ha, that's a good point! We only have autocomplete when the directory exists, but it would make sense to autocomplete when config exists as well (with a console warning). This sounds like a good change to make 👍

**Reply by bholmesdev** [reply]
Also wanted to ask, did you see a warning in your dev console when you configured a collection of the wrong name? We should warn when you configure a collection that doesn't have a matching directory.

**Reply by adamkrupke** [reply]
Uh oh, didn't recall seeing this warning but I do see it now. That's great!

**Reply by bholmesdev** [reply]
Well we wanted to replace that log with something more visible anyways! The fact you missed it is proof for me 🙃
